charon/src/translate_ctx.rs

```diff
@@ -482,6 +492,90 @@ impl<'tcx, 'ctx> TransCtx<'tcx, 'ctx> {
         }
     }
 
+    /// Returns the attributes (`#[...]`) of this item.
+    pub(crate) fn item_attributes(&self, id: DefId) -> &[rustc_ast::Attribute] {
+        use rustc_hir::hir_id::HirId;
+        id.as_local()
+            .map(|local_def_id| self.tcx.hir().attrs(HirId::make_owner(local_def_id)))
+            .unwrap_or_default()
+    }
+
+    /// Translates a rust attribute. Returns `None` if the attribute is a doc comment (rustc
+    /// encodes them as attributes). For now we use `String`s for `Attributes`.
+    pub(crate) fn translate_attribute(&self, attr: &rustc_ast::Attribute) -> Option<Attribute> {
+        use rustc_ast::ast::AttrKind;
+        use rustc_ast_pretty::pprust;
+        match &attr.kind {
+            AttrKind::Normal(normal_attr) => {
+                // Use `pprust` to render the attribute like it is written in the source.
+                use pprust::PrintState;
+                Some(pprust::State::to_string(|s| {
+                    s.print_attr_item(&normal_attr.item, attr.span)
+                }))
+            }
+            AttrKind::DocComment(..) => None,
+        }
+    }
+
+    pub(crate) fn translate_attributes_from_rid(&self, id: DefId) -> Vec<Attribute> {
+        self.item_attributes(id)
+            .iter()
+            .filter_map(|attr| self.translate_attribute(attr))
+            .collect()
+    }
+
+    pub(crate) fn translate_inline_from_rid(&self, id: DefId) -> Option<InlineAttr> {
+        use rustc_attr as rustc;
+        if !self.tcx.def_kind(id).has_codegen_attrs() {
+            return None;
+        }
+        match self.tcx.codegen_fn_attrs(id).inline {
+            rustc::InlineAttr::None => None,
+            rustc::InlineAttr::Hint => Some(InlineAttr::Hint),
+            rustc::InlineAttr::Never => Some(InlineAttr::Never),
+            rustc::InlineAttr::Always => Some(InlineAttr::Always),
+        }
+    }
+
+    pub(crate) fn translate_visibility_from_rid(&self, id: DefId) -> bool {
+        use rustc_hir::def::DefKind::*;
+        match self.tcx.def_kind(id) {
+            AssocConst
+            | AssocFn
+            | Const
+            | Enum
+            | Field
+            | Fn
+            | ForeignTy
+            | Macro { .. }
+            | Mod
+            | Static { .. }
+            | Struct
+            | Trait
+            | TraitAlias
+            | TyAlias
+            | Union
+            | Use => self.tcx.visibility(id).is_public(),
+            // These defs don't have visibility modifiers, which would cause `visibility` to panic.
+            AnonConst
+            | AssocTy
+            | Closure
+            | ConstParam
+            | Ctor { .. }
+            | ExternCrate
+            | ForeignMod
+            | Generator
+            | GlobalAsm
+            | Impl { .. }
+            | ImplTraitPlaceholder
+            | InlineConst
+            | LifetimeParam
+            | OpaqueTy
+            | TyParam
+            | Variant => false,
```

Comment:
Are we sure we want the default to be `false`? I think that given the places where you call `translate_visibility_from_rid` we should simply panic here.

Reply:
That would mean we can't reuse `translate_attributes_from_rid` for e.g. trait impls

Reply:
We could have visibility be `Option<bool>` to indicate missing info

Reply:
I guess the fact that an `impl` should be considered public or not depends on the trait and the type for which it is implemented.
I think we should: 1. add a comment here to explain we need to treat trait impls separately (and explain why), and that we return `false`by default for them; 2. panic for the others; 3. add a comment to the `public` field of `ItemMeta` to explain that in case of trait implementations it has no meaning (and is `false` by default).

Reply:
> I guess the fact that an impl should be considered public or not depends on the trait and the type for which it is implemented (correct?).

In a sense, but we also have non-trait impl blocks and, I just discovered, closures.

Agreed on the steps to take.